core-bundle/src/Resources/contao/drivers/DC_Table.php

```diff
@@ -224,6 +230,14 @@ public function __construct($strTable, $arrModule=array())
 		}
 	}
 
+	public function __destruct()
+	{
+		if (!empty($this->arrSubmit))
+		{
+			throw new \RuntimeException(__CLASS__ . ' must call the submit() method to store data in Contao 5.');
+		}
+	}
```

Comment:
I’m not a big fan of exceptions in destructors, because when this acutally gets executed depends on who holds a reference to `$dc` 😕

How about using something like this 
```php
try {
    return $dc->$act();
} finally {
    $dc->validateSubmit();
}
```
here https://github.com/contao/contao/blob/7b8f4ddcdc957db3f2efcf9c797fd0cbca4200c2/core-bundle/src/Resources/contao/classes/Backend.php#L572

Or maybe there is an even better approach to check this inside `$act()` itself?

Reply:
> Or maybe there is an even better approach to check this inside $act() itself?

That won't work. The whole point is that this method could have been replaced, otherwise it would already work as intended. Adding this to `Backend` would work, but only for the default backend module case, not for anyone calling the DC methods.

Reply:
We cannot throw exceptions in destructors - this interferes with handling of already thrown exceptions during execution + makes exceptions appear at places they were not intended. Including tests for this logic for instance. If you want to enforce a call pattern for an API, this must be done in the methods/accessors. Not calling something and then having no interaction anymore afterwards can't be an error. 😊

Reply:
Well _cannot_ is a hard statement. https://github.com/symfony/http-client/blob/6.1/Response/AsyncResponse.php#L179 can as well, as @Toflar pointed out 😂 
anyway, I don't see another solution that works in all cases, does anyone else?